### PR TITLE
Adding new sites

### PR DESCRIPTION
**Pull request by posiaden123**

### CODE REVIEW

**Comment by Flameish** [comment]
Hey thanks for the pull request! The reason why I haven't added AO3 before is because they already have excellent download options built in and I don't want to create unnecessary traffic for them. 

Also please use [example_com](https://github.com/Flameish/Novel-Grabber/blob/master/src/main/java/grabber/sources/example_com.java) as a template for new sites. The [selectors](https://github.com/posiaden123/Novel-Grabber/blob/de1dcfd27317bcc929bcdd4aed231463460a9f66/src/main/java/grabber/sources/wnmtl_org.java#L88) can create NullPointer  if the elements cant be found (eg. wrong url)

**Comment by posiaden123** [comment]
Hey, thanks for responding! Yeah, I see your point about AO3 and I took it out, but I believe wnmtl should work, cause the NPE is caught getting the chapter list.

File: src/main/java/grabber/sources/wnmtl_org.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class wnmtl_org implements Source{
+    private final String name = "wnmtl.org";
```
Comment (Flameish) [inline src/main/java/grabber/sources/wnmtl_org.java:19]: Website name without domain ending  WNMTL. Whatever they have as their logo basically

File: src/main/java/grabber/sources/wnmtl_org.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    @Override
+    public List<Chapter> getChapterList() {
+        List<Chapter> cList = new ArrayList<>();
+        try {
+            toc = Jsoup.connect(novel.novelLink)
+                    .cookies(novel.cookies)
+                    .userAgent("Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:83.0) Gecko/20100101 Firefox/83.0")
+                    .get();
+            Elements chaps = toc.select("li.wp-manga-chapter > a");
```
Comment (Flameish) [inline src/main/java/grabber/sources/wnmtl_org.java:53]: Chapters are not static and are loaded later  via ajax (POST to admin-ajax.php) see daonovel_com for example

File: src/main/java/grabber/sources/wnmtl_org.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+    @Override
+    public NovelMetadata getMetadata() {
+        NovelMetadata mdata = new NovelMetadata();
+        if (toc != null) {
+            Element title = toc.selectFirst("div.post-title > h1");
+            Element author = toc.selectFirst("div.author-content > a");
+            Element description = toc.selectFirst("div.description-summary > div > p");
+            Element cimg = toc.selectFirst("img.img-responsive");
+            mdata.setTitle(title.text());
```
Comment (Flameish) [inline src/main/java/grabber/sources/wnmtl_org.java:92]: Nullpointer unsafe

File: src/main/java/grabber/sources/n78zaotl_com.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class n78zaotl_com implements Source {
+    private final String name = "78zaotl.com";
```
Comment (Flameish) [inline src/main/java/grabber/sources/n78zaotl_com.java:18]: Name: 78zaoTL

File: src/main/java/grabber/sources/n78zaotl_com.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    private final String url = "https://www.78zaotl.com/xk/";
```
Comment (Flameish) [inline src/main/java/grabber/sources/n78zaotl_com.java:19]: Remove subdirectory

File: src/main/java/grabber/sources/n78zaotl_com.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    @Override
+    public NovelMetadata getMetadata() {
+        NovelMetadata novelMetadata = new NovelMetadata();
+        if (toc != null) {
+            Element title = toc.selectFirst("div.container > h5");
+            //not possible to get description, everything is in one container div
+            //no cover image provided
+            //no author provided
+            novelMetadata.setTitle(title.text());
```
Comment (Flameish) [inline src/main/java/grabber/sources/n78zaotl_com.java:89]: Nullpointer unsafe

File: src/main/java/grabber/sources/n78zaotl_com.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            novelMetadata.setAuthor("");
+            novelMetadata.setDescription("");
```
Comment (Flameish) [inline src/main/java/grabber/sources/n78zaotl_com.java:91]: Synopsis is inside p tags

**Review by Flameish — CHANGES_REQUESTED** [review]
A few small fixes then its good!

File: src/main/java/grabber/sources/wnmtl_org.java (hunk 2 above)
Comment (posiaden123) [reply src/main/java/grabber/sources/wnmtl_org.java:53]: > Chapters are not static and are loaded later via ajax (POST to admin-ajax.php) see daonovel_com for example

Hey, I'm trying to learn more about the entire process! Can you tell me how you managed to figure out the correct way to grab the novel? Sorry if I'm being annoying😅

File: src/main/java/grabber/sources/n78zaotl_com.java (hunk 7 above)
Comment (posiaden123) [reply src/main/java/grabber/sources/n78zaotl_com.java:91]: > Synopsis is inside p tags

could i just select the p tags that come after the container, or do I have to merge them?
I committed what I think works, but let me know!

File: src/main/java/grabber/sources/wnmtl_org.java (hunk 2 above)
Comment (Flameish) [reply src/main/java/grabber/sources/wnmtl_org.java:53]: If you open the Network tab of the Web Developer tools you can see the http requests the browser makes. 
img

As you can see there are no chapters being loaded on first request. Novel-Grabber uses Jsoup to fetch the sites which is static, meaning it cant work with/execute JavaScript (which loads the chapters later on in the browser).

Now you have to look through the requests to find where the chapters are being loaded, in this case there is a POST to admin-ajax.php

img

Next you have to  look at what is being POSTed to the server to replicate it.

img

So to make it work in Novel-Grabber you need to first get the toc page, like normal. The toc object is used to fetch the metadate later on and you need to get the novel id for the ajax request. Then you need to fetch the chapter list via a POST. Jsoup can also do this.

img

Comment (posiaden123) [reply src/main/java/grabber/sources/wnmtl_org.java:53]: Thank you so much! Working on fixes now!

**Comment by posiaden123** [comment]
review when you can

**Comment by Flameish** [comment]
Looks good!

**Pull request merged**